### ISSUE TITLE
Linkerd,kuberenetes: possible memory leak

### ISSUE THREAD

**Issue by Tapanito** [linkerd/linkerd#1343]
From: https://discourse.linkerd.io/t/linkerd-high-memory-usage-on-kubernetes/72/8

## Setup
I'm running Linkerd 1.0.2 as Ingress for Kubernetes cluster I'm using nghttpx as the edge router to route all the traffic to Linkerd, which then distributes the traffic to the necessary recipients.  Linkerd is listening for both http and h2 protocols. The actual configuration of linkerd is this:
```
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: l5d-config
  namespace: engine-stage
data:
  config.yaml: |-
    #  Enable admin panel to monitor liveness and readiness
    admin:
      ip: 0.0.0.0
      port: 9990
    namers:
    - kind: io.l5d.k8s
      experimental: true
    usage:
      enabled: false
    routers:

    - protocol: http
      identifier:
        kind: io.l5d.ingress
        namespace: engine-stage
      servers:
        - port: 8080 
          ip: 0.0.0.0
          clearContext: true
      dtab: /svc => /#/io.l5d.k8s ;
      client:
        kind: io.l5d.global
        loadBalancer:
          kind: ewma
          # number of retries before a node is marked as unavailable
          maxEffort: 10
          decayTimeMs: 10000

    - protocol: h2
      experimental: true
      identifier:
        kind: io.l5d.ingress
        namespace: engine-stage
      servers:
        - port: 8081
          ip: 0.0.0.0
          clearContext: true
      dtab: /svc => /#/io.l5d.k8s ;

---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  labels:
    app: l5d
  name: l5d
  namespace: engine-stage
spec:
  template:
    metadata:
      labels:
        app: l5d
    spec:
      volumes:
      - name: l5d-config
        configMap:
          name: "l5d-config"
      containers:

      - name: l5d
        image: buoyantio/linkerd:1.0.2
        livenessProbe:
          httpGet:
            path: /admin/ping
            port: 9990
            httpHeaders:
          initialDelaySeconds: 5
          periodSeconds: 3
        readinessProbe:
          httpGet:
            path: /admin/ping
            port: 9990
            httpHeaders:
          initialDelaySeconds: 5
          periodSeconds: 5
        env:
        - name: POD_IP
          valueFrom:
            fieldRef:
              fieldPath: status.podIP
        args:
        - /io.buoyant/linkerd/config/config.yaml
        ports:
        - name: http
          containerPort: 8080
          hostPort: 8080
        - name: h2
          hostPort: 8081
          containerPort: 8081
        - name: admin
          hostPort: 9990
          containerPort: 9990
        volumeMounts:
        - name: "l5d-config"
          mountPath: "/io.buoyant/linkerd/config"
          readOnly: 

      - name: kubectl
        image: buoyantio/kubectl:v1.4.0
        args: ["proxy", "-p", "8001"]

---
apiVersion: v1
kind: Service
metadata:
  name: l5d
  namespace: engine-stage
spec:
  selector:
    app: l5d
  ports:
  - name: http
    port: 8080
  - name: h2
    port: 8081
 ```

## Test setup
* Start with a fresh linkerd pod.
* Issue 500gRPC requests per worker, running 10 workers, each worker using a separate connection.

## Results
* Memory usage before serving the requests:
![before](https://cloud.githubusercontent.com/assets/5780819/26397577/abd010e0-406e-11e7-8015-70109a3f4e85.png)
* Memory usage after serving the requests:
![after](https://cloud.githubusercontent.com/assets/5780819/26397767/512aac8a-406f-11e7-8565-3f192f02e9ca.png)

* [Linkerd metrics](https://pastebin.com/1vH1CzpU)

## Issue
As expected, when serving the traffic the linkerd memory increases, however once the traffic stops, the memory usage does not decrease and stays high until the pod is restarted.

Also, linkerd and kubernetes provide two different results for memory usage:
![image](https://cloud.githubusercontent.com/assets/5780819/26397732/35693b42-406f-11e7-8093-63741c7081df.png)

**Comment by adleong** [comment]
Thanks for the report!  Does memory use continue to grow if traffic is left on?  Or does it stabilize at 173MiB?

**Comment by jamessharp** [comment]
Just to let you know - I ran into what looks like a memory leak in production with linkerd 1.0.2 (kubernetes 1.6.2, GKE, daemonsets)

Here's the l5d memory usage for the last week - I upgraded from 1.0.0 to 1.0.2 on Monday afternoon. Performed various restarts in between, downgraded to 1.0.0 again today (Thursday), and everything is looking OK again.

![image](https://cloud.githubusercontent.com/assets/1677854/26449732/ce46b16e-414b-11e7-834e-3379284d33e8.png)

**Comment by olix0r** [comment]
thanks for the great report, @jamessharp!

**Comment by klingerf** [comment]
Hey @jamessharp -- sorry for the delay on this. Can you provide your linkerd config where you're seeing the leak? I can try to reproduce in our test env.

**Comment by jamessharp** [comment]
I've sent @siggy a slack message with the config I've been using. Let me know if there's anything else you need

**Comment by stevej** [comment]
Hi @jamessharp, so far we haven't been able to reproduce this particular issue. Would it be possible to provide a heap dump when linkerd is taking a gig or more memory? We've found heap dumps are a pretty reliable technique for tracking down memory leaks.

Here are two ways to get a heap dump: https://blog.gceasy.io/2015/08/14/how-to-capture-heap-dump-jmap/

Thanks a bunch!

**Comment by adleong** [comment]
Any update here?  @jamessharp are you still seeing this and are you able to capture a heap dump?

**Comment by siggy** [comment]
config file for reference:
```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: l5d-mesh-config
  namespace: linkerd
data:
  config.yaml: |-
    admin:
      port: 9990

    namers:
      - kind: io.l5d.k8s
        experimental: true
        host: localhost
        port: 8001

    routers:
      - protocol: http
        label: http
        dstPrefix: /http
        interpreter:
          kind: io.l5d.namerd
          dst: /$/inet/namerd-sync.linkerd/4100
          namespace: mesh
          transformers:
          - kind: io.l5d.k8s.daemonset
            namespace: linkerd
            port: incoming-http
            service: l5d-mesh
        servers:
        - port: 4143
          ip: 0.0.0.0
        client:
          kind: io.l5d.global
          tls:
            trustCerts:
              - /infraca/ca.crt
            commonName: linkerd

      - protocol: http
        label: api
        dstPrefix: /api
        identifier:
          kind: io.l5d.header
          header: x-ortoo-api-resource
        interpreter:
          kind: io.l5d.namerd
          dst: /$/inet/namerd-sync.linkerd/4100
          namespace: mesh
          transformers:
          - kind: io.l5d.k8s.daemonset
            namespace: linkerd
            port: incoming-http
            service: l5d-mesh
        servers:
        - port: 4144
          ip: 0.0.0.0
        client:
          kind: io.l5d.global
          tls:
            trustCerts:
              - /infraca/ca.crt
            commonName: linkerd

      - protocol: http
        label: incoming-http
        dstPrefix: /
        identifier:
          kind: io.l5d.header
          header: l5d-dst-client
        interpreter:
          kind: io.l5d.namerd
          dst: /$/inet/namerd-sync.linkerd/4100
          namespace: mesh
          transformers:
            - kind: io.l5d.k8s.localnode
        servers:
        - port: 4139
          ip: 0.0.0.0
          tls:
            certPath: /certificates/tls.crt
            keyPath: /certificates/tls.key
---
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  labels:
    app: l5d-mesh
    ortoo.tls.infra: "true"
  name: l5d-mesh
  namespace: linkerd
  annotations:
    ortoo/tls.infra.hosts: '
      *.infra.ortootech.com
      *.default
      linkerd
    '
    ortoo/tls.infra.secret: l5d-mesh-infra-ssl-cert
    ortoo/tls.infra.ca.secret: l5d-infra-ca-cert
spec:
  template:
    metadata:
      labels:
        app: l5d-mesh
    spec:
      volumes:
      - name: l5d-mesh-config
        configMap:
          name: l5d-mesh-config
      - name: l5d-mesh-infra-ssl-cert
        secret:
          secretName: l5d-mesh-infra-ssl-cert
      - name: l5d-infra-ca-cert
        secret:
          secretName: l5d-infra-ca-cert
      containers:
      - name: l5d
        image: buoyantio/linkerd:1.0.0
        env:
        - name: NODE_NAME
          valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
        - name: POD_IP
          valueFrom:
            fieldRef:
              fieldPath: status.podIP
        args:
          - /io.buoyant/linkerd/config/config.yaml
          # - -log.level=DEBUG
        ports:
        - name: incoming-http
          containerPort: 4139
          hostPort: 4139
        - name: http
          containerPort: 4143
          hostPort: 4143
        - name: api
          containerPort: 4144
          hostPort: 4144
        - name: admin
          containerPort: 9990
          hostPort: 9990
        volumeMounts:
        - name: "l5d-mesh-config"
          mountPath: "/io.buoyant/linkerd/config"
          readOnly: true
        - name: l5d-mesh-infra-ssl-cert
          mountPath: /certificates
          readOnly: true
        - name: l5d-infra-ca-cert
          mountPath: /infraca
          readOnly: true

      - name: kubectl
        image: buoyantio/kubectl:v1.6.2
        args:
        - "proxy"
        - "-p"
        - "8001"
```

**Comment by jamessharp** [comment]
Hi Everyone - I've been away so haven't been able to look into this yet.

I've only managed to get it happening in production, which is less than ideal for testing, but with a bit of close monitoring I reckon I can get a memory dump for you

**Comment by wmorgan** [comment]
@jamessharp please do, we would love to dig into this.

**Comment by jamessharp** [comment]
I've had this running all day and got to about 600MB of mem usage. Due to time constraints I'm going to have to take it down. I'm sending the dump over to @siggy now (via slack). Please keep it confidential. Here's the memory usage of the pod over time:

![image](https://user-images.githubusercontent.com/1677854/27494864-a09c51a2-5847-11e7-9b3c-990bd9d02b66.png)

Ignoring the final peak (when I took the dump), it's interesting that kubectl is also consuming more and more memory

**Comment by wmorgan** [comment]
Thanks @jamessharp ! We're on it.

**Comment by siggy** [comment]
@jamessharp That heap dump is consistent with the memory leaks you are seeing. We're trying to get a reproducible environment. Can you tell us a bit about the clients/workers that are connecting to this? Is it a load test tool? Do they timeout/disconnect?

Eclipse Memory Analyzer shows 433,641 `com.twitter.finagle.Stack$Node` objects:

![screen shot 2017-06-26 at 11 51 42 am](https://user-images.githubusercontent.com/236915/27555194-deabcbf6-5a65-11e7-8dfb-a28db0bbf719.png)

Digging into a `com.twitter.finagle.Stack$Node` using `jhat jmap_dump.bin`, and then navigating up via `References to this object`, it appears to reference `scala.collection.immutable.$colon$colon` at least 100 times up the chain.

**Comment by jamessharp** [comment]
@siggy - Glad to see you're making progress!

The setup is our production GKE cluster running roughly 10 nodes, with linkerd on all of them as a service mesh. All traffic was real world traffic - we get ~1m HTTP requests in a day. Requests come in via 2 external facing endpoints that run https://github.com/ortoo/linkerd-proxy (a very simple, but very possibly dodgy, proxy that terminates SSL and strips out l5d headers). The proxies then forward the traffic onto their hosts linkerd and it routes it from there. 

I've now moved those proxies over to use nginx rather than the custom go app, and upgraded to linkerd 1.1.0 and there seems to be no sign of the leak

**Comment by wmorgan** [comment]
Hm, maybe this was fixed in 1.1.0?

OOC, with the advent of `clearContext` (https://linkerd.io/config/1.1.0/linkerd/index.html#server-parameters), do you still need nginx in front?

**Comment by siggy** [comment]
@jamessharp closing this for now as the issue seems to be resolved. please reopen if you see the issue again.
